src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/FrameConnection.cs

```diff
@@ -69,114 +68,141 @@ public void StartRequestProcessing()
 
         private async Task ProcessRequestsAsync()
         {
-            RawStream rawStream = null;
-
             try
             {
-                Task adaptedPipelineTask = Task.CompletedTask;
+                var adaptedPipelineTask = Task.CompletedTask;
+                var input = _context.Input.Reader;
+                var output = _context.Output.Writer;
 
-                if (_connectionAdapters.Count == 0)
-                {
-                    _frame.Input = _context.Input.Reader;
-                    _frame.Output = _context.OutputProducer;
-                }
-                else
+                // Set these before the first await, this is to make sure that we don't yield control
+                // to the transport until we've added the connection to the connection manager
+                _frame.TimeoutControl = this;
+                _context.ServiceContext.ConnectionManager.AddConnection(_context.FrameConnectionId, this);
+                _lastTimestamp = _context.ServiceContext.SystemClock.UtcNow.Ticks;
+
+                if (_connectionAdapters.Count > 0)
                 {
-                    rawStream = new RawStream(_context.Input.Reader, _context.OutputProducer);
+                    var adaptedPipeline = await ApplyConnectionAdaptersAsync();
 
-                    try
+                    if (adaptedPipeline == null)
                     {
-                        var adaptedPipeline = await ApplyConnectionAdaptersAsync(rawStream);
-
-                        _frame.Input = adaptedPipeline.Input.Reader;
-                        _frame.Output = adaptedPipeline.Output;
+                        // We need to assign these anyways because frame.Abort calls via OutputProducer.Abort
+                        _frame.Input = input;
+                        _frame.Output = new OutputProducer(output, _frame, ConnectionId, Log);
```

Comment:
You instantiate `OutputProducer` in the two places were you assign it - create it before the `if` and assign the same instance in both places.